File: backend/app/branches/models/branch_warehouse.py
"""Firm-scoped branch, warehouse, and storage structure persistence models."""

from decimal import Decimal
from typing import ClassVar
from uuid import UUID

from sqlalchemy import (
    JSON,
    Boolean,
    ForeignKey,
    Index,
    Numeric,
    String,
    Text,
    UniqueConstraint,
    and_,
    text,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.business.models import AttributeEntityType, AttributeValueBase
from app.core.database.entity import BaseEntity
from app.core.database.types import UUIDType


class BranchType(BaseEntity):
    """Persist reusable branch type masters per firm."""

    __tablename__ = "branch_types"
    __table_args__ = (
        UniqueConstraint("firm_id", "code", name="UQ_branch_types_firm_code"),
        UniqueConstraint("firm_id", "name", name="UQ_branch_types_firm_name"),
    )

    firm_id: Mapped[UUID] = mapped_column(
        UUIDType(), ForeignKey("firms.id"), nullable=False, index=True
    )
    code: Mapped[str] = mapped_column(String(50), nullable=False)
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    description: Mapped[str | None] = mapped_column(Text)
    is_active: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=True, server_default="true"
    )


class Branch(BaseEntity):
    """Represent one physical operational branch owned by a firm."""

    __tablename__ = "branches"
    __table_args__ = (
        UniqueConstraint("firm_id", "code", name="UQ_branches_firm_code"),
        Index("IX_branches_firm_name", "firm_id", "name"),
        Index("IX_branches_firm_status", "firm_id", "status"),
        # A firm has one default branch. Nothing enforced it, so every branch
        # could carry the flag and "the default" resolved to an arbitrary row.
        # Mirrors UQ_user_firms_active_primary; MySQL ignores the predicate, so
        # BranchWarehouseService stays the authoritative check.
        Index(
            "UQ_branches_default_active",
            "firm_id",
            unique=True,
            postgresql_where=text("is_default AND NOT is_deleted"),
            sqlite_where=text("is_default = 1 AND is_deleted = 0"),
        ),
    )

    firm_id: Mapped[UUID] = mapped_column(
        UUIDType(), ForeignKey("firms.id"), nullable=False, index=True
    )
    code: Mapped[str] = mapped_column(String(50), nullable=False)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    display_name: Mapped[str] = mapped_column(String(200), nullable=False)
    description: Mapped[str | None] = mapped_column(Text)
    business_profile_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("business_profiles.id", ondelete="RESTRICT")
    )
    branch_type_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("branch_types.id", ondelete="RESTRICT")
    )
    branch_manager_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("users.id", ondelete="RESTRICT")
    )
    email: Mapped[str | None] = mapped_column(String(320))
    phone: Mapped[str | None] = mapped_column(String(20))
    mobile: Mapped[str | None] = mapped_column(String(20))
    country_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("geo_countries.id", ondelete="RESTRICT")
    )
    state_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("geo_states.id", ondelete="RESTRICT")
    )
    district_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("geo_districts.id", ondelete="RESTRICT")
    )
    city_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("geo_cities.id", ondelete="RESTRICT")
    )
    postal_code_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("geo_postal_codes.id", ondelete="RESTRICT")
    )
    locality_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("geo_localities.id", ondelete="RESTRICT")
    )
    address_line1: Mapped[str | None] = mapped_column(String(250))
    address_line2: Mapped[str | None] = mapped_column(String(250))
    timezone: Mapped[str | None] = mapped_column(String(100))
    currency_code: Mapped[str | None] = mapped_column(String(3))
    gst_registration: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False, server_default="false"
    )
    pan: Mapped[str | None] = mapped_column(String(32))
    license_number: Mapped[str | None] = mapped_column(String(64))
    working_hours: Mapped[dict[str, object]] = mapped_column(
        JSON, nullable=False, default=dict
    )
    is_default: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False, server_default="false"
    )
    status: Mapped[str] = mapped_column(
        String(20), nullable=False, default="ACTIVE", server_default="ACTIVE"
    )

    warehouses: Mapped[list["Warehouse"]] = relationship(
        back_populates="branch",
        cascade="save-update, merge",
        primaryjoin=lambda: and_(
            Branch.id == Warehouse.branch_id,
            Warehouse.is_deleted.is_(False),
        ),
        lazy="selectin",
        order_by="Warehouse.created_at",
    )


class WarehouseType(BaseEntity):
    """Persist reusable warehouse type masters per firm."""

    __tablename__ = "warehouse_types"
    __table_args__ = (
        UniqueConstraint("firm_id", "code", name="UQ_warehouse_types_firm_code"),
        UniqueConstraint("firm_id", "name", name="UQ_warehouse_types_firm_name"),
    )

    firm_id: Mapped[UUID] = mapped_column(
        UUIDType(), ForeignKey("firms.id"), nullable=False, index=True
    )
    code: Mapped[str] = mapped_column(String(50), nullable=False)
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    description: Mapped[str | None] = mapped_column(Text)
    is_active: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=True, server_default="true"
    )


class Warehouse(BaseEntity):
    """Represent one physical warehouse mapped to a branch."""

    __tablename__ = "warehouses"
    __table_args__ = (
        UniqueConstraint("firm_id", "code", name="UQ_warehouses_firm_code"),
        Index("IX_warehouses_firm_name", "firm_id", "name"),
        Index("IX_warehouses_branch_status", "branch_id", "status"),
        # One default warehouse per branch, for the same reason.
        Index(
            "UQ_warehouses_default_active",
            "branch_id",
            unique=True,
            postgresql_where=text("is_default AND NOT is_deleted"),
            sqlite_where=text("is_default = 1 AND is_deleted = 0"),
        ),
    )

    firm_id: Mapped[UUID] = mapped_column(
        UUIDType(), ForeignKey("firms.id"), nullable=False, index=True
    )
    branch_id: Mapped[UUID] = mapped_column(
        UUIDType(),
        ForeignKey("branches.id", ondelete="RESTRICT"),
        nullable=False,
        index=True,
    )
    code: Mapped[str] = mapped_column(String(50), nullable=False)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    display_name: Mapped[str] = mapped_column(String(200), nullable=False)
    description: Mapped[str | None] = mapped_column(Text)
    warehouse_type_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("warehouse_types.id", ondelete="RESTRICT")
    )
    warehouse_manager_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("users.id", ondelete="RESTRICT")
    )
    business_profile_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("business_profiles.id", ondelete="RESTRICT")
    )
    country_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("geo_countries.id", ondelete="RESTRICT")
    )
    state_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("geo_states.id", ondelete="RESTRICT")
    )
    district_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("geo_districts.id", ondelete="RESTRICT")
    )
    city_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("geo_cities.id", ondelete="RESTRICT")
    )
    postal_code_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("geo_postal_codes.id", ondelete="RESTRICT")
    )
    locality_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("geo_localities.id", ondelete="RESTRICT")
    )
    address_line1: Mapped[str | None] = mapped_column(String(250))
    address_line2: Mapped[str | None] = mapped_column(String(250))
    capacity: Mapped[Decimal | None] = mapped_column(Numeric(18, 3))
    capacity_unit: Mapped[str | None] = mapped_column(String(20))
    is_default: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False, server_default="false"
    )
    temperature_controlled: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False, server_default="false"
    )
    cold_storage: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False, server_default="false"
    )
    hazardous_storage: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False, server_default="false"
    )
    has_receiving_area: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False, server_default="false"
    )
    has_dispatch_area: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False, server_default="false"
    )
    has_returns_area: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False, server_default="false"
    )
    has_inspection_area: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False, server_default="false"
    )
    has_packing_area: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False, server_default="false"
    )
    has_loading_dock: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False, server_default="false"
    )
    status: Mapped[str] = mapped_column(
        String(20), nullable=False, default="ACTIVE", server_default="ACTIVE"
    )

    branch: Mapped[Branch] = relationship(back_populates="warehouses")
    storage_nodes: Mapped[list["WarehouseStorageNode"]] = relationship(
        back_populates="warehouse",
        cascade="save-update, merge",
        primaryjoin=lambda: and_(
            Warehouse.id == WarehouseStorageNode.warehouse_id,
            WarehouseStorageNode.is_deleted.is_(False),
        ),
        lazy="selectin",
        order_by="WarehouseStorageNode.created_at",
    )


class WarehouseStorageNode(BaseEntity):
    """Represent storage hierarchy nodes (area/rack/shelf/bin/receiving)."""

    __tablename__ = "warehouse_storage_nodes"
    __table_args__ = (
        UniqueConstraint(
            "warehouse_id",
            "code",
            name="UQ_warehouse_storage_nodes_warehouse_code",
        ),
        UniqueConstraint(
            "warehouse_id",
            "name",
            "parent_id",
            name="UQ_warehouse_storage_nodes_warehouse_name_parent",
        ),
        Index(
            "IX_warehouse_storage_nodes_warehouse_parent", "warehouse_id", "parent_id"
        ),
        Index("IX_warehouse_storage_nodes_warehouse_type", "warehouse_id", "node_type"),
    )

    warehouse_id: Mapped[UUID] = mapped_column(
        UUIDType(),
        ForeignKey("warehouses.id", ondelete="RESTRICT"),
        nullable=False,
        index=True,
    )
    parent_id: Mapped[UUID | None] = mapped_column(
        UUIDType(),
        ForeignKey("warehouse_storage_nodes.id", ondelete="RESTRICT"),
    )
    node_type: Mapped[str] = mapped_column(String(20), nullable=False)
    code: Mapped[str] = mapped_column(String(50), nullable=False)
    name: Mapped[str] = mapped_column(String(120), nullable=False)
    description: Mapped[str | None] = mapped_column(Text)
    path: Mapped[str] = mapped_column(String(1000), nullable=False)
    sort_order: Mapped[int] = mapped_column(
        nullable=False, default=0, server_default="0"
    )
    is_active: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=True, server_default="true"
    )

    warehouse: Mapped[Warehouse] = relationship(back_populates="storage_nodes")


class BranchAttributeValue(AttributeValueBase):
    """Store one configurable attribute value for a branch."""

    __tablename__ = "branch_attribute_values"
    __table_args__ = (
        UniqueConstraint(
            "branch_id",
            "attribute_definition_id",
            name="UQ_branch_attribute_values_owner_attribute",
        ),
        Index("IX_branch_attribute_values_firm_text", "firm_id", "value_text"),
        Index("IX_branch_attribute_values_firm_number", "firm_id", "value_number"),
        Index("IX_branch_attribute_values_firm_date", "firm_id", "value_date"),
    )

    ENTITY_TYPE: ClassVar[AttributeEntityType] = AttributeEntityType.BRANCH
    OWNER_COLUMN: ClassVar[str] = "branch_id"

    branch_id: Mapped[UUID] = mapped_column(
        UUIDType(),
        ForeignKey("branches.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )


class WarehouseAttributeValue(AttributeValueBase):
    """Store one configurable attribute value for a warehouse."""

    __tablename__ = "warehouse_attribute_values"
    __table_args__ = (
        UniqueConstraint(
            "warehouse_id",
            "attribute_definition_id",
            name="UQ_warehouse_attribute_values_owner_attribute",
        ),
        Index("IX_warehouse_attribute_values_firm_text", "firm_id", "value_text"),
        Index("IX_warehouse_attribute_values_firm_number", "firm_id", "value_number"),
        Index("IX_warehouse_attribute_values_firm_date", "firm_id", "value_date"),
    )

    ENTITY_TYPE: ClassVar[AttributeEntityType] = AttributeEntityType.WAREHOUSE
    OWNER_COLUMN: ClassVar[str] = "warehouse_id"

    warehouse_id: Mapped[UUID] = mapped_column(
        UUIDType(),
        ForeignKey("warehouses.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )

File: desktop/lib/models/branch_warehouse.dart
import 'entities.dart';

class BranchRecord {
  const BranchRecord({
    required this.id,
    this.version = 0,
    required this.firmId,
    required this.code,
    required this.name,
    required this.displayName,
    required this.description,
    required this.branchTypeId,
    required this.branchManagerId,
    required this.businessProfileId,
    required this.email,
    required this.phone,
    required this.mobile,
    this.addressLine1 = '',
    this.addressLine2 = '',
    this.countryId = '',
    this.stateId = '',
    this.districtId = '',
    this.cityId = '',
    this.postalCodeId = '',
    this.localityId = '',
    required this.currencyCode,
    this.gstRegistration = false,
    required this.status,
    required this.isDefault,
    required this.isDeleted,
    required this.warehouseCount,
    required this.createdAt,
  });

  final String id;

  /// The optimistic-concurrency version this record was read at, sent back
  /// as `If-Match` on save so a concurrent edit is refused rather than
  /// silently overwritten. Zero means the server published none, and the
  /// save then carries no precondition.
  final int version;
  final String firmId;
  final String code;
  final String name;
  final String displayName;
  final String description;
  final String branchTypeId;
  final String branchManagerId;
  final String businessProfileId;
  final String email;
  final String phone;
  final String mobile;
  final String addressLine1;
  final String addressLine2;

  /// Where the branch is, as ids into the shared geography masters. There is
  /// no free-text city on this table — these keys are the only way to say it.
  final String countryId;
  final String stateId;
  final String districtId;
  final String cityId;
  final String postalCodeId;
  final String localityId;
  final String currencyCode;
  final bool gstRegistration;
  final String status;
  final bool isDefault;
  final bool isDeleted;
  final int warehouseCount;
  final String createdAt;

  factory BranchRecord.fromJson(Json json) => BranchRecord(
        id: stringValue(json['id']),
        version: (json['version'] as num?)?.toInt() ?? 0,
        firmId: stringValue(json['firm_id']),
        code: stringValue(json['code']),
        name: stringValue(json['name']),
        displayName: stringValue(json['display_name']),
        description: stringValue(json['description']),
        branchTypeId: stringValue(json['branch_type_id']),
        branchManagerId: stringValue(json['branch_manager_id']),
        businessProfileId: stringValue(json['business_profile_id']),
        email: stringValue(json['email']),
        phone: stringValue(json['phone']),
        mobile: stringValue(json['mobile']),
        addressLine1: stringValue(json['address_line1']),
        addressLine2: stringValue(json['address_line2']),
        countryId: stringValue(json['country_id']),
        stateId: stringValue(json['state_id']),
        districtId: stringValue(json['district_id']),
        cityId: stringValue(json['city_id']),
        postalCodeId: stringValue(json['postal_code_id']),
        localityId: stringValue(json['locality_id']),
        currencyCode: stringValue(json['currency_code']),
        gstRegistration: boolValue(json['gst_registration']),
        status: stringValue(json['status']),
        isDefault: boolValue(json['is_default']),
        isDeleted: boolValue(json['is_deleted']),
        warehouseCount: (json['warehouse_count'] as num?)?.toInt() ?? 0,
        createdAt: stringValue(json['created_at']),
      );
}

class WarehouseRecord {
  const WarehouseRecord({
    required this.id,
    this.version = 0,
    required this.firmId,
    required this.branchId,
    required this.code,
    required this.name,
    required this.displayName,
    required this.warehouseTypeId,
    required this.businessProfileId,
    this.addressLine1 = '',
    this.addressLine2 = '',
    this.countryId = '',
    this.stateId = '',
    this.districtId = '',
    this.cityId = '',
    this.postalCodeId = '',
    this.localityId = '',
    required this.capacity,
    required this.capacityUnit,
    required this.status,
    required this.isDefault,
    required this.temperatureControlled,
    required this.coldStorage,
    required this.hazardousStorage,
    this.hasReceivingArea = false,
    this.hasDispatchArea = false,
    this.hasReturnsArea = false,
    this.hasInspectionArea = false,
    this.hasPackingArea = false,
    this.hasLoadingDock = false,
    required this.isDeleted,
    required this.createdAt,
  });

  final String id;

  /// The optimistic-concurrency version this record was read at, sent back
  /// as `If-Match` on save so a concurrent edit is refused rather than
  /// silently overwritten. Zero means the server published none, and the
  /// save then carries no precondition.
  final int version;
  final String firmId;
  final String branchId;
  final String code;
  final String name;
  final String displayName;
  final String warehouseTypeId;
  final String businessProfileId;
  final String addressLine1;
  final String addressLine2;

  /// Where the warehouse is. Same geography masters the branch uses.
  final String countryId;
  final String stateId;
  final String districtId;
  final String cityId;
  final String postalCodeId;
  final String localityId;
  final String capacity;
  final String capacityUnit;
  final String status;
  final bool isDefault;
  final bool temperatureControlled;
  final bool coldStorage;
  final bool hazardousStorage;
  final bool hasReceivingArea;
  final bool hasDispatchArea;
  final bool hasReturnsArea;
  final bool hasInspectionArea;
  final bool hasPackingArea;
  final bool hasLoadingDock;
  final bool isDeleted;
  final String createdAt;

  factory WarehouseRecord.fromJson(Json json) => WarehouseRecord(
        id: stringValue(json['id']),
        version: (json['version'] as num?)?.toInt() ?? 0,
        firmId: stringValue(json['firm_id']),
        branchId: stringValue(json['branch_id']),
        code: stringValue(json['code']),
        name: stringValue(json['name']),
        displayName: stringValue(json['display_name']),
        warehouseTypeId: stringValue(json['warehouse_type_id']),
        businessProfileId: stringValue(json['business_profile_id']),
        addressLine1: stringValue(json['address_line1']),
        addressLine2: stringValue(json['address_line2']),
        countryId: stringValue(json['country_id']),
        stateId: stringValue(json['state_id']),
        districtId: stringValue(json['district_id']),
        cityId: stringValue(json['city_id']),
        postalCodeId: stringValue(json['postal_code_id']),
        localityId: stringValue(json['locality_id']),
        capacity: stringValue(json['capacity']),
        capacityUnit: stringValue(json['capacity_unit']),
        status: stringValue(json['status']),
        isDefault: boolValue(json['is_default']),
        temperatureControlled: boolValue(json['temperature_controlled']),
        coldStorage: boolValue(json['cold_storage']),
        hazardousStorage: boolValue(json['hazardous_storage']),
        hasReceivingArea: boolValue(json['has_receiving_area']),
        hasDispatchArea: boolValue(json['has_dispatch_area']),
        hasReturnsArea: boolValue(json['has_returns_area']),
        hasInspectionArea: boolValue(json['has_inspection_area']),
        hasPackingArea: boolValue(json['has_packing_area']),
        hasLoadingDock: boolValue(json['has_loading_dock']),
        isDeleted: boolValue(json['is_deleted']),
        createdAt: stringValue(json['created_at']),
      );
}

class TypeRecord {
  const TypeRecord({
    required this.id,
    required this.code,
    required this.name,
    required this.description,
    required this.isActive,
    required this.isDeleted,
  });

  final String id;
  final String code;
  final String name;
  final String description;
  final bool isActive;
  final bool isDeleted;

  factory TypeRecord.fromJson(Json json) => TypeRecord(
        id: stringValue(json['id']),
        code: stringValue(json['code']),
        name: stringValue(json['name']),
        description: stringValue(json['description']),
        isActive: boolValue(json['is_active'], fallback: true),
        isDeleted: boolValue(json['is_deleted']),
      );
}

class StorageNodeRecord {
  const StorageNodeRecord({
    required this.id,
    required this.warehouseId,
    required this.parentId,
    required this.nodeType,
    required this.code,
    required this.name,
    required this.path,
    required this.sortOrder,
    required this.isActive,
    required this.isDeleted,
  });

  final String id;
  final String warehouseId;
  final String parentId;
  final String nodeType;
  final String code;
  final String name;
  final String path;
  final int sortOrder;
  final bool isActive;
  final bool isDeleted;

  factory StorageNodeRecord.fromJson(Json json) => StorageNodeRecord(
        id: stringValue(json['id']),
        warehouseId: stringValue(json['warehouse_id']),
        parentId: stringValue(json['parent_id']),
        nodeType: stringValue(json['node_type']),
        code: stringValue(json['code']),
        name: stringValue(json['name']),
        path: stringValue(json['path']),
        sortOrder: (json['sort_order'] as num?)?.toInt() ?? 0,
        isActive: boolValue(json['is_active'], fallback: true),
        isDeleted: boolValue(json['is_deleted']),
      );
}

class BranchQuery {
  const BranchQuery({
    this.status,
    this.branchTypeId,
    this.managerId,
    this.businessProfileId,
    this.cityId,
    this.stateId,
    this.countryId,
    this.includeDeleted = false,
  });

  final String? status;
  final String? branchTypeId;
  final String? managerId;
  final String? businessProfileId;
  final String? cityId;
  final String? stateId;
  final String? countryId;
  final bool includeDeleted;

  Map<String, String> toQuery() => {
        if (status?.isNotEmpty == true) 'status': status!,
        if (branchTypeId?.isNotEmpty == true) 'branch_type_id': branchTypeId!,
        if (managerId?.isNotEmpty == true) 'manager_id': managerId!,
        if (businessProfileId?.isNotEmpty == true)
          'business_profile_id': businessProfileId!,
        if (cityId?.isNotEmpty == true) 'city_id': cityId!,
        if (stateId?.isNotEmpty == true) 'state_id': stateId!,
        if (countryId?.isNotEmpty == true) 'country_id': countryId!,
        if (includeDeleted) 'include_deleted': 'true',
      };
}

class WarehouseQuery {
  const WarehouseQuery({
    this.status,
    this.branchId,
    this.warehouseTypeId,
    this.managerId,
    this.businessProfileId,
    this.cityId,
    this.stateId,
    this.countryId,
    this.includeDeleted = false,
  });

  final String? status;
  final String? branchId;
  final String? warehouseTypeId;
  final String? managerId;
  final String? businessProfileId;
  final String? cityId;
  final String? stateId;
  final String? countryId;
  final bool includeDeleted;

  Map<String, String> toQuery() => {
        if (status?.isNotEmpty == true) 'status': status!,
        if (branchId?.isNotEmpty == true) 'branch_id': branchId!,
        if (warehouseTypeId?.isNotEmpty == true)
          'warehouse_type_id': warehouseTypeId!,
        if (managerId?.isNotEmpty == true) 'manager_id': managerId!,
        if (businessProfileId?.isNotEmpty == true)
          'business_profile_id': businessProfileId!,
        if (cityId?.isNotEmpty == true) 'city_id': cityId!,
        if (stateId?.isNotEmpty == true) 'state_id': stateId!,
        if (countryId?.isNotEmpty == true) 'country_id': countryId!,
        if (includeDeleted) 'include_deleted': 'true',
      };
}
